crates/polars-utils/src/idx_vec.rs

```diff
@@ -178,13 +178,21 @@ impl<T> Drop for UnitVec<T> {
 impl<T> Clone for UnitVec<T> {
     fn clone(&self) -> Self {
         unsafe {
-            let mut me = std::mem::ManuallyDrop::new(Vec::with_capacity(self.len));
-            let buffer = me.as_mut_ptr();
-            std::ptr::copy(self.data_ptr(), buffer, self.len);
-            UnitVec {
-                data: buffer,
-                len: self.len,
-                capacity: NonZeroUsize::new(std::cmp::max(self.len, 1)).unwrap(),
+            if self.capacity.get() == 1 {
+                Self {
+                    data: self.data,
+                    len: self.len,
+                    capacity: self.capacity,
+                }
+            } else {
+                let mut me = std::mem::ManuallyDrop::new(Vec::with_capacity(self.len));
```

Comment:
Another bug I'm afraid, if `self.len == 0` we just allocated a zero-length `Vec` but claim down below that the capacity is 1.

Can you replace this branch with
```rust
let mut copy = Self::with_capacity(self.len);
std::ptr::copy(self.data_ptr(), copy.data_ptr(), self.len);
copy.len = self.len;
```

and then optimize `with_capacity` to not use `reserve` but instead to branch on the requested capacity, if `cap <= 1` it returns `Self::new`, otherwise it makes a `Vec` with the requested capacity, etc.

Reply:
Updated